### PR TITLE
Speed up InferTypes and CInferTypes

### PR DESCRIPTION
**Pull request by jackkoenig**
A couple of minor improvements to performance of InferTypes and CInferTypes.

I've marked this for backporting but I did the deprecation warnings assuming it won't be included in the 1.3.1 release tomorrow.

The reason the traversal was redundant in `DefNode` was that it already mapped on its expression to get the type of the node itself. Due to declaration before use requirements, the node itself cannot be used in it's RHS.

## Benchmarks results

In a normal compilation of FIRRTL, InferTypes is run 6 times and CInferTypes is run once.

### Full compiler run

| Design | Before Runtime (s) | After Runtime (s) | Speedup |
| ---- | ---- | ---- | ----- |
| small | 47.10 ± 0.75 | 45.23 ± 0.33 | 4 % |
| medium | 117.8 ± 0.45 | 108.35 ± 2.5 | 8 % |
| large | 384.9 ± 30.7 | 380.64 ± 8.0 | uncertainty too high |

### JMH Results of LowerTypes

| Design | Before (ms) | After (ms) | Speedup |
| ---- | ---- | ---- | ---- |
| small | 283 ± 9.1 | 178 ± 12.8 | 37 % |
| medium | 698 ± 82.5 | 506 ± 23.6 | 28 % |
| large | 1715 ± 250 |  1290 ± 109 | 25% |

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you update the FIRRTL spec to include every new feature/behavior?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

  - performance improvement

#### API Impact

Deprecates type aliases in `InferTypes` and `CInferTypes`

#### Backend Code Generation Impact

None

#### Desired Merge Strategy

 - Rebase: You will rebase the PR onto master and it will be merged with a merge commit. 

#### Release Notes

Improve performance of InferTypes and CInferTypes

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by albert-magyar** [comment]
> A couple of minor improvements to performance of InferTypes and CInferTypes.
> 
> I've marked this for backporting but I did the deprecation warnings assuming it won't be included in the 1.3.1 release tomorrow.
> 
> The reason the traversal was redundant in `DefNode` was that it already mapped on its expression to get the type of the node itself. Due to declaration before use requirements, the node itself cannot be used in it's RHS.

Makes sense, and several transforms will error or fail to terminate if they are run on a circuit violating this constraint, so `CheckHighForm` is already a true, hard requirement.

File: src/main/scala/firrtl/passes/InferTypes.scala [hunk 1]
```diff
@@ -12,8 +12,11 @@ object InferTypes extends Pass with PreservesAll[Transform] {
 
   override def prerequisites = Dependency(ResolveKinds) +: firrtl.stage.Forms.WorkingIR
 
+  @deprecated("This should never have been public", "1.3.2")
   type TypeMap = collection.mutable.LinkedHashMap[String, Type]
 
+  private type NTypeMap = collection.mutable.HashMap[String, Type]
```
Comment (albert-magyar) [inline src/main/scala/firrtl/passes/InferTypes.scala:18]: Why the 'N'? This holds the types for all declared components, not just nodes.

File: src/main/scala/firrtl/passes/InferTypes.scala [hunk 2]
```diff
@@ -61,7 +64,7 @@ object InferTypes extends Pass with PreservesAll[Transform] {
         val sxx = (sx map infer_types_e(types)).asInstanceOf[DefNode]
         val t = remove_unknowns(sxx.value.tpe)
         types(sx.name) = t
-        sxx map infer_types_e(types)
```
Comment (albert-magyar) [inline src/main/scala/firrtl/passes/InferTypes.scala:64]: That does indeed look redundant.

File: src/main/scala/firrtl/passes/InferTypes.scala (hunk 1 above)
Comment (jackkoenig) [reply src/main/scala/firrtl/passes/InferTypes.scala:18]: `N` was for "new", should I rename it to make it more clear?

**Comment by jackkoenig** [comment]
@albert-magyar should I backport this to `1.2.x` or leave it `1.3.x`?

File: src/main/scala/firrtl/passes/InferTypes.scala (hunk 1 above)
Comment (albert-magyar) [reply src/main/scala/firrtl/passes/InferTypes.scala:18]: I guess? I guess it doesn't really matter.

**Comment by albert-magyar** [comment]
I think 1.2.x. Both changes are just really simple and solid.

File: src/main/scala/firrtl/passes/InferTypes.scala (hunk 1 above)
Comment (jackkoenig) [reply src/main/scala/firrtl/passes/InferTypes.scala:18]: I'll make it `New, it's 2 whole more characters and if it confused you now it'll confuse people later (myself included lol)

Comment (jackkoenig) [reply src/main/scala/firrtl/passes/InferTypes.scala]: I renamed it to `TypeLookup`

**Pull request merged**